common/kv_db.hpp

```diff
@@ -12,16 +12,53 @@
 //    See the License for the specific language governing permissions and
 //    limitations under the License.
 
-#ifndef KV_DB_HPP
-#define KV_DB_HPP
+#ifndef MENDER_COMMON_KV_DB_HPP
+#define MENDER_COMMON_KV_DB_HPP
 
-namespace kv_db {
+#include <common/error.hpp>
+#include <common/expected.hpp>
+#include <common/io.hpp>
+#include <config.h>
 
-class KeyValueDB {
+#include <string>
+#include <mutex>
+#include <vector>
+#include <map>
+#include <functional>
+
+namespace mender::common::kv_db {
+
+using namespace std;
+
+namespace expected = mender::common::expected;
+
+enum ErrorCode {
+	NoError = 0,
+	ParseError,
+	KeyError,
+};
+using Error = mender::common::error::Error<ErrorCode>;
+
+using ExpectedBytes = expected::Expected<vector<byte>, Error>;
+using ExpectedBool = expected::Expected<bool, Error>;
+
+class Transaction {
+public:
+	virtual ~Transaction() {};
+
+	virtual ExpectedBytes Read(const string &key) = 0;
+	virtual ExpectedBool Write(const string &key, const vector<byte> &value) = 0;
+	virtual ExpectedBool Remove(const string &key) = 0;
+};
+
+// Works as a transaction interface as well, which auto-creates a transaction
+// for each operation.
+class KeyValueDataBase : virtual public Transaction, virtual public mender::common::io::Closer {
 public:
-	void hello_world();
+	virtual ExpectedBool WriteTransaction(function<ExpectedBool(Transaction &)> txnFunc) = 0;
+	virtual ExpectedBool ReadTransaction(function<ExpectedBool(Transaction &)> txnFunc) = 0;
```

Comment:
Note that I omitted `OpenRead` and `OpenWrite` from [the original API](https://github.com/mendersoftware/mender/blob/master/store/store.go#L50-L54). While working on this I realized that these are just duplicating what the other functions already do, and the other functions do it better (most notably they auto-commit). I think the only reason they existed is because `ReadTransaction` and `WriteTransaction` did not exist when they were created.

Reply:
Makes perfect sense to me!

Reply:
Worst case scenario we'll add them in later again

Reply:
yup!